### DISCUSSION TITLE
Question on codebase: Isn't `output->usable_area` always scaled?

### DISCUSSION THREAD

**Discussion by tokyo4j** [labwc/labwc#1620] [Q&A]
> [!NOTE]
> This question is just for my curiosity and not directly related to what I'm working on.

What is the purpose of following code in `view_apply_maximized_geometry()` and `output_usable_area_scaled()`?

Looks like it implies the return value of `output_usable_area_in_layout_coords()` might not respect the output scale, but when is it the case? Isn't `output->usable_area` always scaled?

https://github.com/labwc/labwc/blob/53fc8d5e3947ee09015be9e231d998af919a8308/src/view.c#L990-L998

https://github.com/labwc/labwc/blob/53fc8d5e3947ee09015be9e231d998af919a8308/src/output.c#L871-L879

**Comment by johanmalm** [comment]
Good question - and it's been a long time since I looked at this, so might not get this right.

We set `output->usable_area` in `layers.c`. The logic goes `arrange_layers()` --> `arrange_one_layer()` --> `wlr_scene_layer_surface_v1_cofigure()`. And it looks like the last function does not apply scale because it based on `wlr_output_effective_resolution()` (although I might have missed something subtle there and got muddled up).

When `wlr_output_layout_output_coords()` is then called, I think that thinks that takes into account scaling for the x/y coords (at least for the (0,0) position of the output), but it doesn't touch the width/height... and possibly not for the x/y of the usable_area variable (in output co-ordinates). https://github.com/labwc/labwc/blob/53fc8d5e3947ee09015be9e231d998af919a8308/src/output.c#L858C2-L858C33

I have to admit I can't wrap my head around why we would want to do `if (usable.height == output->wlr_output->height ` at the moment.

This might need to `printf()` detective work... and if it's correct as-is, some helpful comment to explain it, and if not, a fix.

**Reply by Consolatis** [reply]
> I have to admit I can't wrap my head around why we would want to do `if (usable.height == output->wlr_output->height ` at the moment.

Me neither.
I have a feeling that this is really old code from the 0.4 / 0.5 era (before scene graph) that was carried around in various forms until today.

>  and if it's correct as-is, some helpful comment to explain it, and if not, a fix.

Agree.

**Reply by tokyo4j** [reply]
Hmm... I still don't get it.

> We set output->usable_area in layers.c. The logic goes arrange_layers() --> arrange_one_layer() --> wlr_scene_layer_surface_v1_cofigure(). And it looks like the last function does not apply scale because it based on wlr_output_effective_resolution() (although I might have missed something subtle there and got muddled up).

`wlr_output_effective_resolution()` seems to return scaled sizes like shown below, so I guess the values set in `output->usable_area` are also scaled. Am I missing something?

```c
void wlr_output_effective_resolution(struct wlr_output *output,
		int *width, int *height) {
	wlr_output_transformed_resolution(output, width, height);
	*width /= output->scale;
	*height /= output->scale;
}
```

I suspect the codes in the description should be removed.

To test this, I managed to force the statement `box.height /= output->wlr_output->scale;` in `view_apply_maximized_geometry` to be executed, by setting the output scale to 0.5 and the height of waybar to 720px (=default height of nested output (720) / scale (0.5) / 2), then windows were not maximized correctly.

![before-remove](https://github.com/labwc/labwc/assets/22029524/8e54a45d-bf15-4e54-a0b2-2b09fcfc233d)

After removing those lines, windows are maximized correctly.
![after-remove](https://github.com/labwc/labwc/assets/22029524/772b0cf9-b5a7-40f5-81d7-3aabe06fb1f2)
